### PR TITLE
Function framework refactor

### PR DESCRIPTION
**Pull request by acquamarin**
This PR refactors the current testing framework:
TableFunction, AggregateFunction and scalarFunction are implemented using the same base class `Function`. All function definitions are stored together in the BuiltInFunction class.

Todos:
1. Binder should have a unified design/interface to match functions.
2. Change paramterTypes/resultType in the function definition from `LogicalTypeID` to `LogicalType`

### CODE REVIEW

File: src/catalog/catalog_content.cpp [hunk 1]
```diff
@@ -204,21 +204,29 @@ void CatalogContent::readFromFile(const std::string& directory, DBFileType dbFil
...
+    } else {
+        // TODO(Ziyi): we should let table function use the same interface to bind.
+        auto funcType = builtInFunctions->getFunctionType(upperCaseName);
+        switch (funcType) {
+        case function::FunctionType::SCALAR:
+            return SCALAR_FUNCTION;
+        case function::FunctionType::AGGREGATE:
+            return AGGREGATE_FUNCTION;
+        default:
+            // LCOV_EXCL_START
```
Comment (andyfengHKU) [inline src/catalog/catalog_content.cpp:220]: move LCOV above default?

File: src/function/built_in_functions.cpp [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+    }
+    case LogicalTypeID::DOUBLE: {
+        return 102;
+    }
+    case LogicalTypeID::INT128:
+    case LogicalTypeID::TIMESTAMP: {
+        return 120;
+    }
+    default: {
+        // LCOV_EXCL_START
```
Comment (andyfengHKU) [inline src/function/built_in_functions.cpp:218]: ditto

File: src/function/built_in_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+    functions.insert({LIST_ANY_VALUE_FUNC_NAME, ListAnyValueFunction::getFunctionSet()});
+}
+
+void BuiltInFunctions::registerStructFunctions() {
+    functions.insert({STRUCT_PACK_FUNC_NAME, StructPackFunctions::getFunctionSet()});
+    functions.insert({STRUCT_EXTRACT_FUNC_NAME, StructExtractFunctions::getFunctionSet()});
+}
+
+void BuiltInFunctions::registerMapFunctions() {
+    functions.insert({MAP_CREATION_FUNC_NAME, MapCreationFunctions::getFunctionSet()});
```
Comment (andyfengHKU) [inline src/function/built_in_functions.cpp:632]: I start thinking maybe we should remove XXX_FUNC_NAME and just write them inside functions. Each function set will expose a getName interface. You can do this in a separate PR.

File: src/function/table_functions/call_functions.cpp [hunk 4]
```diff
@@ -12,6 +12,13 @@ using namespace kuzu::common;
 using namespace kuzu::catalog;
 using namespace kuzu::main;
 
+function_set CurrentSettingFunction::getFunctionSet() {
+    function_set functionSet;
+    functionSet.push_back(
+        std::make_unique<TableFunction>("current_setting", tableFunc, bindFunc, initSharedState));
+    return functionSet;
+}
```
Comment (andyfengHKU) [inline src/function/table_functions/call_functions.cpp:20]: double check coverage

File: src/function/vector_list_functions.cpp [hunk 5]
```diff
@@ -88,721 +88,730 @@ std::unique_ptr<FunctionBindData> ListCreationVectorFunction::bindFunc(
...
-        getExecFunction<timestamp_t>(arguments, vectorFunctionDefinition->execFunc);
+        getExecFunction<timestamp_t>(arguments, scalarFunction->execFunc);
     } break;
     case LogicalTypeID::INTERVAL: {
-        getExecFunction<interval_t>(arguments, vectorFunctionDefinition->execFunc);
+        getExecFunction<interval_t>(arguments, scalarFunction->execFunc);
     } break;
     default: {
-        throw NotImplementedException("ListSortVectorFunction::bindFunc");
+        throw NotImplementedException("ListSortFunction::bindFunc");
```
Comment (andyfengHKU) [inline src/function/vector_list_functions.cpp:423]: LCOV

Comment (andyfengHKU) [reply src/function/vector_list_functions.cpp:423]: Actually just check the coverage report across this PR.

File: src/include/parser/expression/parsed_function_expression.h [hunk 6]
```diff
@@ -8,28 +8,29 @@ namespace parser {
 class ParsedFunctionExpression : public ParsedExpression {
 public:
     ParsedFunctionExpression(std::string functionName, std::string rawName, bool isDistinct = false)
-        : ParsedExpression{common::FUNCTION, std::move(rawName)}, isDistinct{isDistinct},
```
Comment (andyfengHKU) [inline src/include/parser/expression/parsed_function_expression.h:11]: Can u keep this expression type FUNCTION. I think for parsed expression there is no way to differentiate whether it's scalar or aggregate

**Review by andyfengHKU — APPROVED** [review]

File: src/function/table_functions/call_functions.cpp (hunk 4 above)
Comment (acquamarin) [reply src/function/table_functions/call_functions.cpp:20]: I think it is a false positive. They are guaranteed to be covered
